### PR TITLE
CI update

### PR DESCRIPTION
**Pull request by jason-ntu**
## Summary

Update CI

## Details

1. Use a customized container with circom installed to reduce repetitive installation.
2. Add `build-and-test.yml` as a reusable workflow to simplify the codes in `main-ci.yml` (named as `CI.yml` orignally)
3. Update script in package.json and the Lint job in the workflow so that it should pass now.

## Impacted Areas

`./.github/workflows/`
`./package.json`

## Tests

Some settings for Container Registry (i.e., image on github) and PAT are required (see [ref](https://stackoverflow.com/questions/64033686/how-can-i-use-private-docker-image-in-github-actions)).
The related Dockerfile for the Container Registry is appended in `./.github/workflows/`.

### CODE REVIEW

**Comment by ryanycw** [comment]
Hi, could you please provide more information about the current CICD process? I noticed there are three PRs related to CICD. Are reviews required for previous PRs?

**Comment by ryanycw** [comment]
Update: the purpose of this PR is to finish #23

**Comment by ryanycw** [comment]
I haven't looked into it yet. And just to keep a note that the difference of "Build, Test and Lint /" and "CI /" might be confusing.

**Comment by ryanycw** [comment]
@jason-ntu Is this PR ready to be under reviewed? Or do you have other fixes you want to change first.

**Comment by jason-ntu** [comment]
@ryanycw It's done now. I have remove the "CI" and left only "Build, Test and Lint" and "Build and Test".

**Comment by jason-ntu** [comment]
Note that some potential issues might be found only after this branch is merged (and thus triggered in the later PR), like the issue related to PAT, container or some other env settings. So please notify me if you have merged it 🙏

**Comment by ryanycw** [comment]
> Note that some potential issues might be found only after this branch is merged (and thus triggered in the later PR), like the issue related to PAT, container or some other env settings. So please notify me if you have merged it 🙏

Hi, I would like to ask why not push the latest commit without [skip ci]. I thought that would trigger the actions, and we can determine if the CICD flow is working as expected.

**Comment by jason-ntu** [comment]
> > Note that some potential issues might be found only after this branch is merged (and thus triggered in the later PR), like the issue related to PAT, container or some other env settings. So please notify me if you have merged it 🙏
> 
> Hi, I would like to ask why not push the latest commit without [skip ci]. I thought that would trigger the actions, and we can determine if the CICD flow is working as expected.

@ryanycw this PR triggers the old CICD (written in social-tw's fork) and the new CICD (written in jason-ntu's fork), and both will fail. So if the latest commit is not skipped then we need to find a way to "force merge" it.

Even if we can do the force merging, we still cannot determine if the (new) CICD flow is working as expected, since it must not.

Specifically, the new CICD consists of two workflows `main-ci.yml` and `build-and-test.yml`. `main-ci.yml` can only call `build-and-test.yml` successfully if `build-and-test.yml` is already merged, so we still need to merge it first. I got this conclusion through some experiments, but the environment is not the same so I am not 100% sure. What I am certain about is that for now the `main-ci.yml` cannot find `build-and-test.yml` in this PR, but it did in the experimental PRs in my fork.

**Comment by jason-ntu** [comment]
> > > Note that some potential issues might be found only after this branch is merged (and thus triggered in the later PR), like the issue related to PAT, container or some other env settings. So please notify me if you have merged it 🙏
> > 
> > 
> > Hi, I would like to ask why not push the latest commit without [skip ci]. I thought that would trigger the actions, and we can determine if the CICD flow is working as expected.
> 
> And according to #23, can we have an isolated check for the build so it would be easier to identify whether the issue is in tests or build?

I think we can already do that by checking the result in each job...? Or maybe I miss something please correct me.

image

**Comment by jason-ntu** [comment]
> > > > Note that some potential issues might be found only after this branch is merged (and thus triggered in the later PR), like the issue related to PAT, container or some other env settings. So please notify me if you have merged it 🙏
> > > 
> > > 
> > > Hi, I would like to ask why not push the latest commit without [skip ci]. I thought that would trigger the actions, and we can determine if the CICD flow is working as expected.
> > 
> > 
> > And according to #23, can we have an isolated check for the build so it would be easier to identify whether the issue is in tests or build?
> 
> Yes we can. It's a choice between "saving resource & code simplicity" and "shortening the checking time". The additional build result cannot be pipelined to other jobs as in circleci. (Actually it can, by using [actions/upload-artifact](https://github.com/actions/upload-artifact), which takes about 1 hour to upload and 15-30 to download.)

Should I add it or not?

**Comment by ryanycw** [comment]
> Yes we can. It's a choice between "saving resource & code simplicity" and "shortening the checking time". The additional build result cannot be pipelined to other jobs as in circleci. (Actually it can, by using [actions/upload-artifact](https://github.com/actions/upload-artifact), which takes about 1 hour to upload and 15-30 to download.)

Ahhh, I see. Thank you for clearing up the concerns. In my personal opinion, I am good with keeping the isolated build out since we can find the issue by digging in the steps of the actions. If we decided to go with CircleCI in the future, I would prefer we put the isolated build in, since not all contributor has the access to look into CircleCI logs, thus, I believe should be harder for them to identify the issue. I think, for now everything looks great! Thank you for the great work and explanation. 

I think this PR is good to merge, would anyone double-check on this before we merge? @vivianjeng @NOOMA-42 @Lisooo790926 just mentioned some names that I have in mind who might have experience with this, so if anyone else has any great ideas on this PR, feel free to review them. Hopefully, we can get this PR pass before afternoon so that other tests can use this to ensure the result is correct.

**Comment by ryanycw** [comment]
Another minor issue I want to ask for your experience is, if we did this PR in the branch of this Repo instead of your fork, will the actions be activated in this PR, or is it the same and require us to merge to see the result of the actions? @jason-ntu

**Comment by jason-ntu** [comment]
> Another minor issue I want to ask for your experience is, if we did this PR in the branch of this Repo instead of your fork, will the actions be activated in this PR, or is it the same and require us to merge to see the result of the actions? @jason-ntu

@ryanycw  Yeah I just tried a similar PR (with merely different container and user name) in my fork and it acted as I had expected.

image

**Comment by jason-ntu** [comment]
> > Another minor issue I want to ask for your experience is, if we did this PR in the branch of this Repo instead of your fork, will the actions be activated in this PR, or is it the same and require us to merge to see the result of the actions? @jason-ntu
> 
> @ryanycw Yeah if we did this PR in the branch of this Repo instead of your fork, will the actions be activated in this PR
I just tried a similar PR (with merely different container and user name) in my fork and it acted as I had expected.
> 
> image

**Comment by jason-ntu** [comment]
BTW I just found that I forgot to delete the old `CI.yml` (again) 😂
I think that's the reason why the old CI was activated.
Now I've deleted it in the latest commit.

File: .github/workflows/main-ci.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+              uses: actions/checkout@v3
+            - name: Install node.js
+              uses: actions/setup-node@v3
+              with:
+                  node-version: '18'
+            - name: Install packages
+              run: yarn
+            - name: Lint
+              run: |
+                  yarn lint:fix
```
[the remark this answers is not in the document]
Comment (ryanycw) [reply .github/workflows/main-ci.yml:43]: I am not sure if that is going to cause any issues. If anyone have concerns in this. I would prefer we remove it. What do you think @jason-ntu?

[the remark this answers is not in the document]
Comment (jason-ntu) [reply .github/workflows/main-ci.yml:43]: Oh I know nothing aboot those command. Just copy and paste from here image

@vivianjeng

[the remark this answers is not in the document]
Comment (jason-ntu) [reply .github/workflows/main-ci.yml:43]: @vivianjeng Do I understand it right? Or how should I modify it?

[the remark this answers is not in the document]
Comment (ryanycw) [reply .github/workflows/main-ci.yml:43]: So in the official Unirep Social, they only did the check, what about we remove the fix part? And if there is a future need, we put it back. 

https://github.com/Unirep/Unirep-Social/blob/414ed475268daa93996bb367efb8b2e9a6524eee/.circleci/config.yml#L45

[the remark this answers is not in the document]
Comment (jason-ntu) [reply .github/workflows/main-ci.yml]: OK it's done~

**Pull request merged**